rs/execution_environment/src/canister_manager.rs

```diff
@@ -2042,15 +2043,41 @@ impl CanisterManager {
                 );
             }
             Some(snapshot) => {
-                // Verify the provided snapshot id belongs to this canister.
-                if snapshot.canister_id() != canister_id {
-                    return (
-                        Err(CanisterManagerError::CanisterSnapshotInvalidOwnership {
-                            canister_id,
-                            snapshot_id,
-                        }),
-                        NumInstructions::new(0),
-                    );
+                // Verify the provided `snapshot_id` belongs to a canister controlled by the sender.
+                // Only perform the check if target canister to load the snapshot
+                // is not the same as the one owning the snapshot.
+                let snapshot_canister_id = snapshot.canister_id();
+                if snapshot_canister_id != canister_id {
+                    match state.canister_state(&snapshot_canister_id) {
+                        None => {
+                            // The below case should never happen as if the snapshot still exists, it
+                            // should be associated with an existing canister. If it happens, it indicates
+                            // a bug, so log an error message for investigation.
+                            snapshot_exists_without_associated_canister.inc();
+                            error!(
+                                self.log,
+                                "[EXC-BUG]: Canister {} does not exist although there's a snapshot {} associated with it.",
+                                snapshot_canister_id,
+                                snapshot_id,
+                            );
+                            return (
+                                Err(CanisterManagerError::CanisterNotFound(snapshot_canister_id)),
```

Comment:
This reveals the snapshot canister ID that we wanted to hide, but since this error is not supposed to happen (it could only happen if there's a bug in the implementation), I believe it's fine.

Reply:
Yeah I'm not sure what else we could here besides returning a completely unrelated error.